src/storage/src/batch_table/storage_table.rs

```diff
@@ -0,0 +1,598 @@
+            .collect_vec();
+        let keyspace = Keyspace::table_root(store, &table_id);
+        Self {
+            keyspace,
+            table_columns,
+            schema,
+            pk_serializer,
+            mapping,
+            pk_indices,
+            dist_key_indices,
+            dist_key_in_pk_indices,
+            vnodes,
+            table_option,
+        }
+    }
+
+    pub fn schema(&self) -> &Schema {
+        &self.schema
+    }
+}
+
+/// Point get
+impl<S: StateStore> StorageTable<S> {
+    /// Check whether the given `vnode` is set in the `vnodes` of this table.
+    fn check_vnode_is_set(&self, vnode: VirtualNode) {
+        let is_set = self.vnodes.is_set(vnode as usize).unwrap();
+        assert!(
+            is_set,
+            "vnode {} should not be accessed by this table: {:#?}, dist key {:?}",
+            vnode, self.table_columns, self.dist_key_indices
+        );
+    }
+
+    /// Get vnode value with `indices` on the given `row`. Should not be used directly.
+    fn compute_vnode(&self, row: &Row, indices: &[usize]) -> VirtualNode {
+        let vnode = if indices.is_empty() {
+            DEFAULT_VNODE
+        } else {
+            row.hash_by_indices(indices, &CRC32FastBuilder {})
+                .to_vnode()
+        };
+
+        tracing::trace!(target: "events::storage::storage_table", "compute vnode: {:?} key {:?} => {}", row, indices, vnode);
+
+        // FIXME: temporary workaround for local agg, may not needed after we have a vnode builder
+        if !indices.is_empty() {
+            self.check_vnode_is_set(vnode);
+        }
+        vnode
+    }
+
+    /// Get vnode value with given primary key.
+    fn compute_vnode_by_pk(&self, pk: &Row) -> VirtualNode {
+        self.compute_vnode(pk, &self.dist_key_in_pk_indices)
+    }
+
+    /// Try getting vnode value with given primary key prefix, used for `vnode_hint` in iterators.
+    /// Return `None` if the provided columns are not enough.
+    fn try_compute_vnode_by_pk_prefix(&self, pk_prefix: &Row) -> Option<VirtualNode> {
+        self.dist_key_in_pk_indices
+            .iter()
+            .all(|&d| d < pk_prefix.0.len())
+            .then(|| self.compute_vnode(pk_prefix, &self.dist_key_in_pk_indices))
+    }
+
+    /// `vnode | pk`
+    fn serialize_pk_with_vnode(&self, pk: &Row) -> Vec<u8> {
+        let mut output = Vec::new();
+        output.put_slice(&self.compute_vnode_by_pk(pk).to_be_bytes());
+        self.pk_serializer.serialize(pk, &mut output);
+        output
+    }
+
+    /// Get a single row by point get
+    pub async fn get_row(&mut self, pk: &Row, epoch: u64) -> StorageResult<Option<Row>> {
+        let serialized_pk = self.serialize_pk_with_vnode(pk);
+        let read_options = self.get_read_option(epoch);
+        assert!(pk.size() <= self.pk_indices.len());
+        let key_indices = (0..pk.size())
+            .into_iter()
+            .map(|index| self.pk_indices[index])
+            .collect_vec();
```

Comment:
Can we disable `bloom_filter` in batch mode? Thus we don't need the `pk_indices` in `StorageTable`. cc @st1page @Li0k

Reply:
maybe we can disable it temporarily because for the batch query it is difficult to hit the bloom filter? cc @fuyufjh @hzxa21

Reply:
If the batch query performs scan/lookup on distribution key, we can still turn on bloom_filter. I remembered we have a logic to decide whether the pk/prefix matches the distribution key and turn on bloom filter for batch query. @Li0k Can you double check?

Reply:
> If the batch query performs scan/lookup on distribution key, we can still turn on bloom_filter. I remembered we have a logic to decide whether the pk/prefix matches the distribution key and turn on bloom filter for batch query. @Li0k Can you double check?

yes, now we check `dist_key_indics` and `key_indics`  if it match the `read_pattern` for checking `bloom_filter`. So it rely on `If the batch query performs scan/lookup on distribution key`.  But i am not sure the `bloom_filter` hit rate for batch query request